packages/fxa-auth-server/lib/payments/capability.ts

```diff
@@ -229,6 +245,115 @@ export class CapabilityService {
     ];
   }
 
+  /**
+   * Determine the subscription eligibility path for a user for a given plan,
+   * considering existing IAP subscriptions in the process.
+   * Will throw an error if the targetPlanId does not match with a known plan
+   */
+  public async getPlanEligibility(
+    uid: string,
+    targetPlanId: string,
+    useFirestoreProductConfigs: boolean = false
+  ) {
+    const allPlans = await this.stripeHelper.allAbbrevPlans();
+
+    // Create a map of planId: abbrevPlan for speed/ease of lookup later without iterating
+    const allPlansByPlanId: { [key: string]: AbbrevPlan } = allPlans.reduce(
+      (acc, plan) => {
+        return {
+          ...acc,
+          [plan.plan_id]: plan,
+        };
+      },
+      {}
+    );
+
+    const targetPlan = allPlansByPlanId[targetPlanId];
+
+    if (!targetPlan) throw error.unknownSubscriptionPlan(targetPlanId);
+
+    const { productSet: targetProductSet } = productUpgradeFromProductConfig(
+      targetPlan,
+      useFirestoreProductConfigs
+    );
+
+    if (!targetProductSet) {
+      return SubscriptionEligibilityResult.INVALID;
+    }
+
+    // Fetch all user's subscriptions from all sources
+    const stripeSubscriptions = await this.fetchSubscribedPricesFromStripe(uid);
+    const stripeSubscribedPlans = stripeSubscriptions
+      .map((planId) => allPlansByPlanId[planId])
+      .filter((plan) => plan);
+    const appleIapSubscriptions = await this.fetchSubscribedPricesFromAppStore(
+      uid
+    );
+    const appleIapSubscribedPlans = appleIapSubscriptions
+      .map((planId) => allPlansByPlanId[planId])
+      .filter((plan) => plan);
+    const playIapSubscriptions = await this.fetchSubscribedPricesFromPlay(uid);
+    const playIapSubscribedPlans = playIapSubscriptions
+      .map((planId) => allPlansByPlanId[planId])
+      .filter((plan) => plan);
+
+    const iapSubscribedPlans = [
+      ...appleIapSubscribedPlans,
+      ...playIapSubscribedPlans,
+    ];
+
+    // Lookup whether user holds an IAP subscription with a shared product set to the target
+    const iapRoadblock = iapSubscribedPlans.some((abbrevPlan) => {
+      const { productSet } = productUpgradeFromProductConfig(
+        abbrevPlan,
+        useFirestoreProductConfigs
+      );
+
+      return productSet?.some((name) => targetProductSet.includes(name));
+    });
+
+    // Users with an IAP subscription to the productSet that we're trying to subscribe
+    // to should not be allowed to proceed
+    if (iapRoadblock) {
```

Comment:
How easy would it be to split the iapRoadblock logic into it's own function?

Reason I'm asking is we'll probably have to update the logic where we currently use`getSubscriptionUpdateEligibility` so that it also does the IAP checks.

I don't think we need to do it as part of this ticket. Do you know if there is a follow up ticket for that? Might need to create one otherwise if necessary.

Reply:
Spoke about this in person with @StaberindeZA but documenting here for historical reasons.

`getSubscriptionUpdateEligibility` currently doesn't have the information needed to make a decision WRT IAP subs, since IAP roadblock needs to consider all subscriptions under a given user rather than two individual plans. Adding IAP roadblock to the `getSubscriptionUpdateEligibility` util would fairly significantly increase it's size and complexity.

I believe the path forward here is to use `getPlanEligibility` in all locations that currently use `getSubscriptionUpdateEligibility`, and move `getSubscriptionUpdateEligibility` out of fxa-shared since it's currently used by the frontend. Meghan is already working on the latter part of that, and I'll create a ticket for the former.